piecrust/src/imports.rs

```diff
@@ -248,13 +249,20 @@ pub(crate) fn c(
         let arg = &arg_buf[..arg_len as usize];
 
         callee.write_argument(arg);
+        callee.clear_eco_mode();
         let ret_len = callee
             .call(name, arg.len() as u32, callee_limit)
             .map_err(Error::normalize)?;
         check_arg(callee, ret_len as u32)?;
 
         // copy back result
         callee.read_argument(&mut memory[argbuf_ofs..][..ret_len as usize]);
+        callee.read_eco_mode_buf(&mut memory[eco_mode_ofs..][..ECO_MODE_LEN]);
+        // if depth of the stack is 3 or more, economic mode is not read
+        const ECO_STACK_DEPTH: usize = 2;
+        if env.nth_from_top(ECO_STACK_DEPTH).is_some() {
+            callee.clear_eco_mode();
+        }
```

Comment:
I really don't think this should be the mechanism. While I don't think contracts should be able to make decisions for others, I also don't think we should stop contracts from reading what a callee puts in their buffer.

This feels like a bandaid for a deeper issue. I suggest either having a double buffer - one for writing by the contract and reading by the host, and another for reading by the contract and writing by the host - or make it easy for contracts to *always* specify the economic mode - by modifying the `wrap_call` functions.

Reply:
I strongly disagree with this characterization of the problem. The issue exists whether we consider downstream or not. The bottom line is that a callee should not be able to set the callers economic mode, regardless of call stack height. As it stands this is exactly what happens, but we make an exception for call stack heights larger than 2, which is why I call this a bandaid.

A double buffer would totally solve the problem. You could even just double the size of the buffer and treat each half differently - which would be my favorite approach. Expanding:

```text
                     1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2

Caller
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|          Write Buffer          |           Read Buffer          |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
                                                  |
                +------->-------->-------->--------
Callee          |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|          Write Buffer          |           Read Buffer          |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

The contract writes to the write buffer, and reads from the read buffer, while the host reads from the write buffer, and writes to the read buffer. This means that any callee will effectively only ever set the read buffer of the caller, and as such the caller's decision can never be influence by the callee. This, together with zeroing the contract's write buffer - ensuring the previous decisions don't influence the current one - would serve perfectly well as the economic mode mechanism.

I also think it is a bad design decision to couple `piecrust` with downstream's chosen protocol call stack height. We might as well merge this repository with `rusk` at that point.

Reply:
I am also very interested in keeping Piecrust architecturally sound and agnostic to such peculiarities, yet I need a scheme that works. As far as I can see, it is not possible to come up with a scheme which will not take call stack height into account (or some kind of exception for height 1). This is a derivative of having transfer contract (as a contract). Other solution would be to move the EP logic into host, but I think this would be even more inconsistent with our overall design.

Reply:
Also, I do not think this is as bad as being equivalent of merging `Piecrust` into `rusk`. We need a quite general parameter to Piecrust which would mean: `number of layers of privileged contracts`. In our case it would be `1`, as we have one layer of privileged contracts, i.e., a transfer contract, which is calling all other contracts. And Piecrust would have some buffers accessible only to privileged contracts, that is basically the idea.

Reply:
No, the scheme does not collapse in this case. The transfer contract is always free to - and likely will - write what it read from the "read buffer" to its own "write buffer" so it can be picked up by the host if necessary. It simply becomes a positive, conscious (i.e. by its code), decision, instead of having been perhaps inadvertently set by a contract that it happened to call. 

BTW in the diagram the callee is at the bottom, not at the top, and as such the transfer contract would be the top contract. Reference frame problem :smile: 

While I agree with you that keeping this is not as bad as merging the whole repository into `rusk`, it is a very significant coupling that I think we should avoid. To understand how significant it is, one just has to think of how (if it was merged) the proxy contract PR would have had an influenced `piecrust`, had this already been merged.

Reply:
I see that the difference between our approaches is that I think in terms of what is possible (or allowed), and you think in terms of what can be consciously done (or not easily). If the intention is to make it less accidental but possible, then your scheme will work. I wanted to have a scheme which does not allow callees beyond 2nd level to set the EP arguments at all. Yet if we assume they can do it and the only assumption is that the info is consciously read by callers - the scheme you proposed is doable. Glad this is settled - I will implement it and let you know.